java/src/main/java/com/google/oak/transparency/EndorsementVerifier.java

```diff
@@ -96,51 +87,46 @@ public Result<Boolean, Exception> verifyRekorSignature(
    * the {@code body}. This public is separately checked against a known and
    * trusted (root) public key.
    *
-   * @param body
-   * @param contentBytes
-   * @return an empty Result if the verification succeeds, or an Exception wrapped
-   *         in a Result otherwise
+   * @param body         The body of the Rekor log entry
+   * @param contentBytes The serialized content
+   * @return Empty if the verification succeeds, or an error message otherwise
```

Comment:
Semantically this does not seem accurate -- it seems unidiomatic to expect that a method returns an empty optional to signal success. OTOH I am also not super happy with our own custom Result class. Maybe we should just give in and use exceptions like most other Java code? :roll_eyes:

Reply:
When going with exceptions, it would need to be a checked one:

void verify(...) throws VerificationFailedException {}

This is OK but the catching is cumbersome.

The other option I can think of is to return Optional<Error>, where Error is mostly a string. Would this address your concern?

Reply:
Yeah, an empty optional to indicate success is definitely counterintuitive. I think idiomatic Java would indeed to use an exception (or at last it was idiomatic when I last wrote Java 15 years ago). `Optional<Error>` would make it better, as at least the Optional would work the way you expect.

Reply:
Going with `Optional<Error>`. Personally, I wouldn't mind `Optional<String>` as long as the javadoc is clear about the semantics.

After thinking some more, I don't believe that this is a use case for exceptions - a failing signature verification is not "exceptional". `void verify(...) throws VerificationFailedException {}` is really bad: A function signalling to the caller *only* via exceptions, no other effects or side effects. In addition, it wastes (leaves unused) the valuable return slot.